File: sorting/src/main/java/com/justintullgren/hackerrank/sorting/BubbleSort.java
package com.justintullgren.hackerrank.sorting;

import java.util.Collections;
import java.util.List;

public class BubbleSort {
    public static <T extends Comparable<T>> void sort(List<T> list) {
        boolean isSorted = false;
        int length = list.size() - 1;
        T a, b;
        while (!isSorted) {
            isSorted = true;
            for (int i = 0; i < length; i++) {
                a = list.get(i);
                b = list.get(i + 1);
                if (a.compareTo(b) > 0) {
                    Collections.swap(list, i, i + 1);
                    isSorted = false;
                }

            }
            length--;
        }
    }
}

File: sorting/src/main/kotlin/com/justintullgren/hackerrank/sorting/BubbleSort.kt
package com.justintullgren.hackerrank.sorting

fun countSwaps(a: Array<Int>): Unit {

    val answer = bubbleSort(a)

    println("Array is sorted in ${answer.first} swaps.")
    println("First Element: ${answer.second.first()}")
    println("Last Element: ${answer.second.last()}")

}

fun bubbleSort(array: Array<Int>): Pair<Int, Array<Int>> {
    var swaps = 0
    var isSorted = false
    var last = array.lastIndex - 1

    while (!isSorted) {
        isSorted = true
        for (i in 0..last) {
            val next = i + 1
            if (array[i] > array[next]) {
                val temp = array[i]
                array[i] = array[next]
                array[next] = temp
                isSorted = false
                swaps++
            }
        }
        last--
    }
    return Pair(swaps, array)
}

